### PR TITLE
Add the Interactivity API runtime to WooCommerce blocks

### PR DESCRIPTION
**Pull request by DAreRodz**
## What?

Includes the current [WP directives runtime](https://github.com/WordPress/block-hydration-experiments/issues/80) inside the WooCommerce blocks repository and makes it ready to use.

Tracking issue: https://github.com/woocommerce/woocommerce/issues/42486

## Why?

This is required to start testing the Interactivity API in WooCommerce blocks until it is natively available in WordPress.

## How?

- Copy-pasting the WP directives runtime code.
- Updating the Webpack configuration to compile the directive's JSX code into Preact.
- Adding an experimental filter to enable directives hydration (`false` by default).

## Testing instructions

See https://github.com/woocommerce/woocommerce-blocks/pull/8026#issuecomment-1382153094

* [x] Do not include in the Testing Notes

### CODE REVIEW

File: woocommerce-gutenberg-products-block.php
Comment (luisherranz) [inline woocommerce-gutenberg-products-block.php]: We don't need the filter or the enqueue because `wp_enqueue_script( 'wp-directives-runtime' )` should always be called as [a dependency of another script](https://github.com/woocommerce/woocommerce-blocks/blob/748a1e9a1f12166ca014bfc4472cf5167d916e30/src/BlockTypes/SimplePriceFilter.php#L39).

File: woocommerce-gutenberg-products-block.php
Comment (luisherranz) [inline woocommerce-gutenberg-products-block.php]: We can get rid of this too.

File: tsconfig.base.json
Comment (gigitux) [inline tsconfig.base.json]:
```suggestion
			"@woocommerce/interactivity": [ "assets/js/base/interactivity" ],
```

What do you think if we rename it to @woocommerce/interactivity?

File: bin/webpack-configs.js
Comment (gigitux) [inline bin/webpack-configs.js:846]:
```suggestion
								presets: [
									[
										'@wordpress/babel-preset-default',
										{
											modules: false,
											targets: {
												browsers: [
													'extends @wordpress/browserslist-config',
												],
											},
										},
									],
									[
										'@babel/preset-react',
										{
											runtime: 'automatic',
											importSource: 'preact',
										},
									],
								],
```

It seems that it is not necessary to install and use `@babel/plugin-proposal-optional-chaining`. We can use `@wordpress/babel-preset-default`.

File: woocommerce-gutenberg-products-block.php
Comment (gigitux) [inline woocommerce-gutenberg-products-block.php:308]:
```suggestion
	wp_enqueue_script( 'wp-directives-runtime' );
```

The script `wp-directives-runtime` isn't enqueued.

**Review by gigitux — CHANGES_REQUESTED** [review]

**Comment by sunyatasattva** [comment]
Great job! It looks amazing and makes us super excited!

@gigitux and me were testing the functionality and there seems to be a bug with the implementation of the router. Relative URLs are apparently a problem when fetching.

You can see this in the Pagination component if, instead of clicking on “Next Page” you click on the numbers for the page on a Product Query block set without the option “Inherit query from template”.

For some reason (which should probably be addressed in Gutenberg), the pagination component renders a relative URL in the `href` of the numbers, but an absolute URL in the `href` of the “Next” and “Previous” page. The interactivity API seems to be unable to parse the relative URLs as relatives, as can be seen in the network tab.

Is this a known issue? Should the interactivity only work with absolute URLs?

Screenshot 2023-01-19 at 17 16 32

Screenshot 2023-01-19 at 17 17 26

https://d.pr/v/MaBLdc

File: woocommerce-gutenberg-products-block.php
Comment (sunyatasattva) [reply woocommerce-gutenberg-products-block.php]: We just saw these comments after commenting ourselves on the enqueueing. Your proposal for removal of the filter makes sense, so sorry @DAreRodz for the confusion. We removed the redundant comments 😅 

However, as of the removal of these, the testing instructions for this PR are not correct anymore. So we should update them so that testers can try out these specific changes 😊

File: tsconfig.base.json
Comment (DAreRodz) [reply tsconfig.base.json]: That's a good suggestion. 👍

File: bin/webpack-configs.js
Comment (DAreRodz) [reply bin/webpack-configs.js:846]: I have to say that I already tried, and `@wordpress/babel-preset-default` was messing up with Preact, adding `@wordpress/element` instead, so I opted to use that plugin instead of the preset you mentioned.

This is the part that replaces `createElement` with `@wordpress/element`:

https://github.com/WordPress/gutenberg/blob/trunk/packages/babel-preset-default/index.js#L80-L82

**Comment by DAreRodz** [comment]
Hey @sunyatasattva, thank you very much for your review. 😄 

Glad you discovered that bug; we were not aware of it. 😅 It should be a quick fix, though, so I'll open a PR right away at https://github.com/WordPress/block-hydration-experiments/ and update the code here with the changes.

**Comment by DAreRodz** [comment]
@sunyatasattva, @gigitux, this should be ready. ☝️

File: woocommerce-gutenberg-products-block.php
Comment (luisherranz) [inline woocommerce-gutenberg-products-block.php]: I don't think `wp_enqueue_scripts` is the correct action here because if other people use it to enqueue these scripts their code may run it before this registration. 

I'd rather use another action that runs before this one (instead of lowering the priority).

Comment (DAreRodz) [reply woocommerce-gutenberg-products-block.php]: I would say the proper action is `init`, considering these examples I've found in this repo.

All the functions below call `wp_register_script`.

https://github.com/woocommerce/woocommerce-blocks/blob/a69fc12bd55dc49ad4d43e48715731e1f7e13241/src/Assets/AssetDataRegistry.php#L69

https://github.com/woocommerce/woocommerce-blocks/blob/79e7d221110c3714eb37589982549f87614cc5ab/src/AssetsController.php#L37

https://github.com/woocommerce/woocommerce-blocks/blob/c29c25d64a83aa7acc43c368a18aafe39ea058ec/src/Domain/Services/GoogleAnalytics.php#L32

Comment (DAreRodz) [reply woocommerce-gutenberg-products-block.php]: Changed `wp_enqueue_scripts` to `init` in https://github.com/woocommerce/woocommerce-blocks/pull/8220/commits/05ca8abc8d19bd71ffee6afecccb24994d977e95.

Comment (luisherranz) [reply woocommerce-gutenberg-products-block.php]: Perfect. Thanks, David.

**Review by luisherranz — APPROVED** [review]
Codewise it looks fine to me. We have to rename everything to `woo-`, but we can do that in a different PR.

I'm going to approve it, but please wait until someone from Woo has also approved it to merge it.

**Comment by luisherranz** [comment]
Also, @darerodz: 
- Do you have any idea why some E2E tests are failing? 
- Can you run `npm run lint` scripts locally and see if you can't reproduce why the JS and MD linting actions are failing?

**Comment by gigitux** [comment]
Thanks for updating the PR. Tomorrow I will take a look!
> Also, @DAreRodz:
> 
> * Do you have any idea why some E2E tests are failing?
> * Can you run `npm run lint` scripts locally and see if you can't reproduce why the JS and MD linting actions are failing?

1. We have some issues with E2E tests. I'm working on fixing them. So it is expected that they fail.
2. Same about the linter error about MD.

**Comment by DAreRodz** [comment]
> * Can you run `npm run lint` scripts locally and see if you can't reproduce why the JS and MD linting actions are failing?

I've just run locally `npm run lint:js:report` and `npm run lint:md:docs` (the ones that correspond to the failing GH actions), but I haven't seen any error. 🤔

I'm restarting both actions just in case.

**Comment by DAreRodz** [comment]
Oh, I just saw your comment, @gigitux! 😄 Thanks for the clarification.

**Review by gigitux — APPROVED** [review]
LGTM! Thanks a lot for your work! Amazing 🚀 
I added the label `skip-changelog`, and edited the description adding a checkbox about the testing instructions. This way, the release manager will know that he doesn't add this PR in the testing instruction.

**Review by sunyatasattva — APPROVED** [review]
Exciting!

**Comment by DAreRodz** [comment]
Thanks, folks! 😊 

I think I don't have permission to merge this PR apart from enabling auto-merge―which doesn't work because there are some failing checks. It would be great if you could merge and close this PR for me. 🙏

cc: @gigitux @sunyatasattva

**Comment by gigitux** [comment]
@DAreRodz, thanks for your work! Merged!

**Pull request merged**
